### ISSUE TITLE
Lightweight deletes using subqueries significantly slower in 24.3

### ISSUE THREAD

**Issue by the4thamigo-uk** [ClickHouse/ClickHouse#66892]
**Describe the unexpected behaviour**

In 24.3, lightweight deletions of the following form appear to run much slower in 24.3 vs 24.2.

```
DELETE FROM tbl
WHERE (a,b,c) in (
  select a,b,c from another_tbl
)
```

The subquery returns very quickly (typically only a few hundred rows), but adding the `DELETE` causes the drop in performance compared to 24.2.

After the CH upgrade we go from few seconds to >1min for the mutation to complete.

We have found a workaround that speeds up the mutation significantly, but is less than ideal : 

```
DELETE FROM tbl
WHERE concatWithSeparator('|', a,b,c) in (
  select concatWithSeparator('|', a,b,c) from another_tbl
)
```

* Both tables have `ORDER BY a,b,c` and do not have partitions specified.
* `a` is a time dimension, where values are in _hourly_ granularity
* `tbl` contains about a million rows for each hour
* the subquery always returns values for deletion for a single hour (the rows _are not_ distributed over all time).
* as mentioned above the subquery returns a small number of rows to delete, up to a few hundred at most
* if the subquery returns no rows, the statement completes quickly, so it looks like it is slow only when there is one or more rows to delete
* deleting a single row takes about the same time as deleting a few hundred.

**Comment by SaltTan** [comment]
maybe related? 
* https://github.com/ClickHouse/ClickHouse/issues/65299

**Comment by SaltTan** [comment]
Could you check what was happening during the mutation in `system.part_log`?

**Comment by the4thamigo-uk** [comment]
@SaltTan I am going to try to setup a reproducible example I think. This is a serious issue for us as we rely on this type of query. Also it is [used in dbt clickhouse](https://github.com/ClickHouse/dbt-clickhouse/blob/main/dbt/include/clickhouse/macros/materializations/incremental/incremental.sql#L228C99-L229) which is also a concern for us.

**Comment by the4thamigo-uk** [comment]
@SaltTan As luck would have it, we have experienced the issue today so I can give you the `parts_log`. 

The delete statement that seems to have caused an issue is initiated at about 14:06:54. As you can see from the log we have a lot of `RemovePart` lines then we have the very slow `MutatePart`, `MergeParts` with high memory usage and accompanying OOM exceptions.

[parts_log.csv.gz](https://github.com/user-attachments/files/16501344/parts_log.csv.gz)

**Comment by CurtizJ** [comment]
Should be fixed in #68052. However it's strange that wrapping with `concatWithSeparator` speeds up the query.

**Comment by the4thamigo-uk** [comment]
@CurtizJ Thanks for this. We will revert our workaround and see if the issue is resolved for us.
